packages/iTwinUI-react/src/core/Tree/Tree.tsx

```diff
@@ -224,40 +244,178 @@ export const Tree = <T,>(props: TreeProps<T>) => {
+            indexInGroup: node.indexInGroup,
+            parentNodeId: node.parentNode?.nodeProps.nodeId,
+            scrollToParent: node.parentNode
+              ? () => {
+                  const parentNodeId = node.parentNode?.nodeProps.nodeId;
+                  const parentNodeIndex = flatNodesList.findIndex(
+                    (n) => n.nodeProps.nodeId === parentNodeId,
+                  );
+                  setScrollToIndex(parentNodeIndex);
+                }
+              : undefined,
           }}
         >
-          {nodeRenderer(flatNode.nodeProps)}
+          {nodeRenderer(node.nodeProps)}
         </TreeContext.Provider>
-      ))}
-    </ul>
+      );
+    },
+    [firstLevelNodesList.length, flatNodesList, nodeRenderer],
+  );
+
+  const [scrollToIndex, setScrollToIndex] = React.useState<number>();
+  const flatNodesListRef = React.useRef<FlatNode<T>[]>(flatNodesList);
+  flatNodesListRef.current = flatNodesList;
+
+  React.useEffect(() => {
+    setTimeout(() => {
+      if (scrollToIndex !== undefined) {
+        const nodeId = flatNodesListRef.current[scrollToIndex].nodeProps.nodeId;
+        const nodeElement = treeRef.current?.ownerDocument.querySelector(
+          `#${nodeId}`,
+        ) as HTMLElement;
+        nodeElement?.focus();
+        // Need to reset that if navigating with mouse and keyboard,
+        // e.g. going to the previous node with keyboard
+        // and then clicking the next node with mouse.
+        setScrollToIndex(undefined);
+      }
+    });
+  }, [scrollToIndex]);
+
+  const handleFocus = (event: React.FocusEvent) => {
+    if (treeRef.current?.contains(event.relatedTarget)) {
+      return;
+    }
+
+    const items = getFocusableNodes();
+    if (items.length > 0) {
+      items[focusedIndex.current]?.focus();
+    }
+  };
+
+  return (
+    <>
+      {enableVirtualization ? (
+        <VirtualizedTree
+          flatNodesList={flatNodesList}
+          itemRenderer={itemRenderer}
+          scrollToIndex={scrollToIndex}
+          onFocus={handleFocus}
+          onKeyDown={handleKeyDown}
+          ref={treeRef}
+          className={className}
+          style={style}
+          {...rest}
+        />
+      ) : (
+        <TreeElement
+          onKeyDown={handleKeyDown}
+          onFocus={handleFocus}
+          className={className}
+          style={style}
+          ref={treeRef}
+          {...rest}
+        >
+          {flatNodesList.map((_, i) => itemRenderer(i))}
+        </TreeElement>
+      )}
+    </>
   );
 };
 
+const TreeElement = React.forwardRef(
+  (
+    {
+      children,
+      className,
+      ...rest
+    }: {
+      children: React.ReactNode;
+      onKeyDown: React.KeyboardEventHandler<HTMLUListElement>;
+      onFocus: React.FocusEventHandler<HTMLUListElement>;
+      className?: string;
+      style?: React.CSSProperties;
+      id?: string;
+    },
+    ref: React.ForwardedRef<HTMLUListElement>,
```

Comment:
oh, didnt know you can define types like that too :D
Usually we put those as generics using `<>`